### PR TITLE
Pinning main

### PR DESCRIPTION
**Pull request by rwails**
Adds a CPU-pinning option to Shadow's main branch. The feature is optionally enabled with a CLI switch.

### CODE REVIEW

**Comment by RWails** [comment]
This is failing CI on errors outside of any modules changed. Is that OK?

**Comment by robgjansen** [comment]
Ugg. When did the deprecated `g_memmove` become an error? I guess I should fix that first, merge in here, and re-run the CI tests.

**Comment by stevenengler** [comment]
> Ugg. When did the deprecated `g_memmove` become an error? I guess I should fix that first, merge in here, and re-run the CI tests.

This might be caused by GiHub changing the `ubuntu-latest` test runner from Ubuntu 18.04 to 20.04. You can probably just set the runner to `ubuntu-18.04` and leave the deprecation fix for a future PR.

https://github.com/shadow/shadow/blob/5e7b83ca1ac61f7a8706964f8321fb60ccc668f5/.github/workflows/build_shadow.yml#L15-L17

**Review by robgjansen — APPROVED** [review]
Overall this looks great, thanks!

I just had two questions that I'm unsure about, which means comments in the code for those items would be helpful (though not required) even if you don't end up changing the code.

File: src/main/host/affinity.c [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+    for (size_t idx = 0; idx < _global_platform_info.n_cpus; ++idx) {
+        CPUInfo *p_info = &_global_platform_info.p_cpus[idx];
+        _global_platform_info.max_cpu_num =
+            MAX(_global_platform_info.max_cpu_num, p_info->logical_cpu_num);
+    }
+
+    if (rc) {
+        error("Could not run `lscpu`, which is required for CPU pinning.");
+        return -1;
+    }
```
[the remark this answers is not in the document]
Comment (RWails) [reply src/main/host/affinity.c]: fixed

File: src/main/host/affinity.c [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+    assert(_global_platform_info.n_cpus > 0);
+
+    const CPUInfo *p_cpu_info = &_global_platform_info.p_cpus[0];
+
+    for (size_t idx = 0; idx < _global_platform_info.n_cpus; ++idx) {
+        const CPUInfo *rhs = &_global_platform_info.p_cpus[idx];
+        if (_cpuinfo_compare(p_cpu_info, rhs) == 1) {
+            // In this case, rhs is preferred to lhs
+            p_cpu_info = rhs;
```
[the remark this answers is not in the document]
Comment (RWails) [reply src/main/host/affinity.c:145]: nope! The relation forms a total ordering, so only one pass is required to compute the minimum (in other words, most preferred) CPU to schedule next

**Pull request merged**